org/mozilla/jss/ssl/javax/JSSEngine.java

```diff
@@ -0,0 +1,1296 @@
+package org.mozilla.jss.ssl.javax;
+
+import java.lang.*;
+import java.util.*;
+
+import java.nio.ByteBuffer;
+
+import javax.net.ssl.*;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.mozilla.jss.nss.*;
+import org.mozilla.jss.ssl.*;
+import org.mozilla.jss.pkcs11.*;
+import org.mozilla.jss.provider.javax.crypto.*;
+
+public class JSSEngine extends javax.net.ssl.SSLEngine {
+    /*
+     * TODO list:
+     *
+     *  - Allow client authentication.
+     *
+     * Optional list:
+     *
+     *  - SSLSession object interactions?
+     */
+
+    public static Logger logger = LoggerFactory.getLogger(JSSEngine.class);
+
+    private static int BUFFER_SIZE = 1 << 12;
+
+    private boolean is_client = false;
+    private String peer_info = null;
+    private String hostname = null;
+    private BufferProxy read_buf = null;
+    private BufferProxy write_buf = null;
+    private SSLFDProxy ssl_fd = null;
+
+    private PK11Cert cert = null;
+    private PK11PrivKey key = null;
+    private X509KeyManager[] key_managers = null;
+    private X509TrustManager[] trust_managers = null;
+
+    private boolean need_client_auth = false;
+    private boolean want_client_auth = false;
+
+    private SSLEngineResult.HandshakeStatus handshake_state = SSLEngineResult.HandshakeStatus.NOT_HANDSHAKING;
+
+    private SSLCipher[] enabled_ciphers = null;
+    private SSLVersion min_protocol = null;
+    private SSLVersion max_protocol = null;
+
+    private JSSSession session = null;
+    private int unknown_state_count = 0;
+
+    private boolean step_handshake = false;
+
+    private boolean is_outbound_closed = false;
+    private boolean is_inbound_closed = false;
+
+    private SSLException ssl_exception = null;
+    private boolean seen_exception = false;
+
+    public JSSEngine() {
+        super();
+
+        peer_info = "";
+        session = new JSSSession(BUFFER_SIZE);
+        logger.debug("JSSEngine: constructor()");
+    }
+
+    public JSSEngine(String peerHost, int peerPort) {
+        super(peerHost, peerPort);
+
+        peer_info = peerHost + ":" + peerPort;
+        session = new JSSSession(BUFFER_SIZE);
+        session.setPeerHost(peerHost);
+        session.setPeerPort(peerPort);
+        logger.debug("JSSEngine: constructor(" + peerHost + ", " + peerPort + ")");
+    }
+
+    public JSSEngine(String peerHost, int peerPort, PK11Cert localCert, PK11PrivKey localKey) {
```

Comment:
Just an aside: this constructor is almost never called. It is only called by people who explicitly construct an instance of SSLEngine, instead of asking for one from the JSS Provider's `SSLContext` instance.